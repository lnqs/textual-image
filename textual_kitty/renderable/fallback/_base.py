"""Provides functionality to render images with Rich as unicode characters."""

import logging
from pathlib import Path
from typing import override

from PIL import Image as PILImage

from textual_kitty.geometry import Size
from textual_kitty.renderable._base import _Base
from textual_kitty.terminal import TerminalSizeInformation

logger = logging.getLogger(__name__)


class _FallbackBase(_Base):
    """Rich renderable to render images in terminal.

    This Renderable renders an image as unicode characters. Used as fallback if other protocols are not available.
    """

    @override
    def __init__(self, image: str | Path | PILImage.Image, width: int | None = None) -> None:
        super().__init__(image, width)
        self._prepared_image: PILImage.Image | None = None

    @override
    def is_prepared(self, size: Size, term_size: TerminalSizeInformation) -> bool:
        if self._prepared_image and self._prepared_image.size == self._calculate_image_size_for_render_size(
            size, term_size
        ):
            return True
        return False

    @override
    def prepare(self, size: Size, term_size: TerminalSizeInformation) -> None:
        logger.debug("Scaling fallback image for rendering")
        self._prepared_image = self._prepare_image(size, term_size)

    @override
    async def prepare_async(self, size: Size, term_size: TerminalSizeInformation) -> None:
        logger.debug("Scaling fallback image for rendering asynchrounouly")
        self._prepared_image = await self._run_in_thread(lambda: self._prepare_image(size, term_size))

    def _prepare_image(self, size: Size, term_size: TerminalSizeInformation) -> PILImage.Image:
        raise NotImplementedError()

    def _calculate_image_size_for_render_size(self, render_size: Size, term_size: TerminalSizeInformation) -> Size:
        raise NotImplementedError()
