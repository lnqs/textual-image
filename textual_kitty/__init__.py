"""Package to render images in Rich and Textual.

Provides Rich Renderables and a Textual Widget to render images via the
Terminal Graphics Protocol (<https://sw.kovidgoyal.net/kitty/graphics-protocol/>),
allowing compatible terminals (i.e. Kitty) to display image data as well as a unicode
fallback.
"""
