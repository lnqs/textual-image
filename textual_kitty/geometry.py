"""Geometry data structures."""

from typing import NamedTuple


class Size(NamedTuple):
    """The dimensions (width and height) of a rectangular region."""

    width: int
    """The width."""
    height: int
    """The height."""
