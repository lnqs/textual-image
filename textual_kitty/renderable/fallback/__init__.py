"""Rich `Renderable`s to display images as unicode characters.

To be used if no other protocols are available
"""
