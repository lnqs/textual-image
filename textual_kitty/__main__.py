"""Run the textual_kitty demo."""

from textual_kitty import demo

demo.run()
