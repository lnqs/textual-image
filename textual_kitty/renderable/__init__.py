"""rich `Renderable`s to display images in terminal."""

import logging
import sys
from typing import Type

from textual_kitty.renderable import tgp
from textual_kitty.renderable.fallback import colored, grayscale

logger = logging.getLogger(__name__)

Image: Type[tgp.Image | colored.Image | grayscale.Image]

is_tty = sys.__stdout__ and sys.__stdout__.isatty()

if is_tty and tgp.query_terminal_support():
    logger.debug("Terminal Graphics Protocol support detected")
    Image = tgp.Image
elif is_tty:
    logger.debug("No Terminal Graphics Protocol support detected, falling back to colored unicode")
    Image = colored.Image
else:
    logger.debug("Not connected to a terminal, falling back to grayscale unicode")
    Image = grayscale.Image
