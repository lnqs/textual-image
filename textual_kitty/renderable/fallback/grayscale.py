"""Provides functionality to render images with Rich as unicode characters."""

from contextlib import nullcontext
from typing import cast, override

from PIL import Image as PILImage
from rich.console import RenderResult
from rich.segment import Segment

from textual_kitty.geometry import Size
from textual_kitty.renderable.fallback._base import _FallbackBase
from textual_kitty.terminal import TerminalSizeInformation

CHARACTERS = [
    "█",  # FULL BLOCK
    "▓",  # DARK SHADE
    "▒",  # MEDIUM SHADE
    "░",  # LIGHT SHADE
    " ",  # SPACE
]

CHARACTER_LOOKUP = {int(255 / len(CHARACTERS) * i): c for i, c in enumerate(CHARACTERS)}


class Image(_FallbackBase):
    """Rich renderable to render images in terminal.

    This Renderable renders an image as unicode characters. Used as fallback if other protocols are not available.
    """

    @override
    def _prepare_image(self, size: Size, term_size: TerminalSizeInformation) -> PILImage.Image:
        with nullcontext(self.image) if isinstance(self.image, PILImage.Image) else PILImage.open(self.image) as image:  # type: ignore
            image = image.resize(self._calculate_image_size_for_render_size(size, term_size))
            image = image.convert("L")
            return cast(PILImage.Image, image)

    @override
    def _calculate_image_size_for_render_size(self, render_size: Size, term_size: TerminalSizeInformation) -> Size:
        return Size(
            int(render_size.width), int(render_size.height / 2 / (term_size.cell_width / term_size.cell_height))
        )

    @override
    def _render(self) -> RenderResult:
        if not self._prepared_image:
            return

        for y in range(0, self._prepared_image.height):
            for x in range(self._prepared_image.width):
                character = self._get_pixel(x, y)
                yield Segment(character)
            yield Segment("\n")

    def _get_pixel(self, x: int, y: int) -> str:
        if not self._prepared_image:
            return " "

        pixel = cast(int, self._prepared_image.getpixel((x, y)))
        brightness = 255 - pixel
        index = int(255 / len(CHARACTER_LOOKUP)) * int(brightness / int(255 / len(CHARACTER_LOOKUP)))
        return CHARACTER_LOOKUP[index]
