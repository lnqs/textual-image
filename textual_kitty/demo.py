#!/usr/bin/env python

"""Showcase textual-kitty Renderables/Widgets."""

from argparse import ArgumentParser
from pathlib import Path
from typing import Type, cast

from PIL import Image as PILImage
from PIL import ImageOps

from textual_kitty.renderable._base import _Base

TEST_IMAGE = Path(__file__).parent / "gracehopper.jpg"


def demo_rich(Image: Type[_Base]) -> None:
    """Showcase textual-kitty's Rich Renderable for images."""
    from rich.console import Console
    from rich.style import Style
    from rich.table import Table

    console = Console()

    console.print("KittyImage from path:", style=Style(bold=True))
    console.print(Image(TEST_IMAGE))

    console.print("\nKittyImage from pillow image:", style=Style(bold=True))
    with PILImage.open(TEST_IMAGE) as image:
        console.print(Image(image))

    console.print("\nSame KittyImage instance rendered twice:", style=Style(bold=True))
    image = Image(TEST_IMAGE)
    console.print(image)
    console.print(image)

    console.print("\nKittyImage with width set to 10:", style=Style(bold=True))
    console.print(Image(TEST_IMAGE, width=10))

    console.print("\nKittyImage inside table:", style=Style(bold=True))
    table = Table("Name", "Birthday", "Picture")
    table.add_row("Grace Hopper", "December 9, 1906", Image(TEST_IMAGE))
    console.print(table)


def demo_textual(Renderable: Type[_Base]) -> None:
    """Showcase textual-kitty's Textual Widget for images."""
    from textual import on
    from textual.app import App, ComposeResult
    from textual.containers import Container, HorizontalScroll, VerticalScroll
    from textual.widgets import Button, Collapsible, Footer, Header

    from textual_kitty.widget import Image

    class DemoApp(App[None]):
        DEFAULT_CSS = """
        DemoApp {
            #content {
                layout: grid;
                grid-size: 5 2;
            }

            #content > * {
                border: round gray;
                align: center middle;
            }

            .small {
                width: 6;
            }

            .full-row {
                column-span: 5;
            }

            .border {
                border: round white;
            }

            .width-auto {
                width: auto;
            }

            .width-full {
                width: 100%;
            }
        }
        """

        def compose(self) -> ComposeResult:
            yield Header()
            with VerticalScroll(id="content"):
                with Container() as c:
                    c.border_title = "Grace from path"
                    yield Image(TEST_IMAGE, image_renderable_type=Renderable)
                with Container() as c:
                    c.border_title = "Grace from PIL image"
                    with PILImage.open(TEST_IMAGE) as image:
                        yield Image(image, image_renderable_type=Renderable)
                with Container() as c:
                    c.border_title = "Collapsible Grace"
                    with Collapsible(collapsed=False):
                        yield Image(TEST_IMAGE, image_renderable_type=Renderable)
                with Container() as c:
                    c.border_title = "Small Grace"
                    yield Image(TEST_IMAGE, image_renderable_type=Renderable, classes="small")
                with Container() as c:
                    c.border_title = "Flippable Grace"
                    with PILImage.open(TEST_IMAGE) as image:
                        yield Image(image, image_renderable_type=Renderable, id="flippable")
                    yield Button("Flip", classes="width-full")
                with HorizontalScroll(classes="full-row") as c:
                    c.border_title = "Many Graces (huge, loaded async)"
                    with PILImage.open(TEST_IMAGE) as image:
                        image = image.resize((image.width * 4, image.height * 4))
                    for _ in range(50):
                        yield Image(
                            image, image_renderable_type=Renderable, classes="border width-auto", load_async=True
                        )
            yield Footer()

        @on(Button.Pressed)
        def button_pressed(self, _event: Button.Pressed) -> None:
            widget = self.query_one("#flippable", Image)
            widget.image = ImageOps.flip(cast(PILImage.Image, widget.image))

    DemoApp().run()


def run() -> None:
    """Run the textual_kitty demo."""
    parser = ArgumentParser(description="Demo the capabilities of textual-kitty")
    parser.add_argument("mode", choices=["rich", "textual"], nargs="?", default="rich")
    parser.add_argument(
        "-p", "--protocol", choices=["auto", "tgp", "colored-fallback", "grayscale-fallback"], default="auto"
    )
    arguments = parser.parse_args()

    if arguments.protocol == "auto":
        from textual_kitty.renderable import Image
    elif arguments.protocol == "tgp":
        from textual_kitty.renderable.tgp import Image
    elif arguments.protocol == "colored-fallback":
        from textual_kitty.renderable.fallback.colored import Image
    elif arguments.protocol == "grayscale-fallback":
        from textual_kitty.renderable.fallback.grayscale import Image

    if arguments.mode == "rich":
        demo_rich(Image)
    elif arguments.mode == "textual":
        demo_textual(Image)


if __name__ == "__main__":
    run()
