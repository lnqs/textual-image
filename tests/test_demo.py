import io
from typing import Any, AsyncContextManager, cast
from unittest.mock import patch

from rich.console import Console
from textual.app import App
from textual.pilot import Pilot
from textual.widgets import Button

from textual_kitty import demo
from textual_kitty.renderable.fallback.grayscale import Image


def test_arg_parsing() -> None:
    with patch("sys.argv", ["demo", "rich"]):
        with patch("textual_kitty.demo.demo_rich") as demo_rich:
            demo.run()
    assert demo_rich.called

    with patch("sys.argv", ["demo", "textual"]):
        with patch("textual_kitty.demo.demo_textual") as demo_textual:
            demo.run()
    assert demo_textual.called

    with patch("sys.argv", ["demo", "rich", "-p", "tgp"]):
        with patch("textual_kitty.demo.demo_rich") as demo_rich:
            demo.run()
    assert demo_rich.call_args[0][0].__module__ == "textual_kitty.renderable.tgp"

    with patch("sys.argv", ["demo", "rich", "-p", "colored-fallback"]):
        with patch("textual_kitty.demo.demo_rich") as demo_rich:
            demo.run()
    assert demo_rich.call_args[0][0].__module__ == "textual_kitty.renderable.fallback.colored"

    with patch("sys.argv", ["demo", "rich", "-p", "grayscale-fallback"]):
        with patch("textual_kitty.demo.demo_rich") as demo_rich:
            demo.run()
    assert demo_rich.call_args[0][0].__module__ == "textual_kitty.renderable.fallback.grayscale"


def test_rich() -> None:
    stdout = io.StringIO()
    with patch.object(Console, "file", stdout):
        demo.demo_rich(Image)
    assert "░░░░▒▒░░▒▒" in stdout.getvalue()


async def test_textual() -> None:
    # This is incredibly hacky. But is seems to work.
    awaitable = None

    def run_wrapper(app: App[Any]) -> None:
        nonlocal awaitable
        awaitable = app.run_test()

    # This doesn't actually test much, but at least we run the code.
    with patch.object(App, "run", run_wrapper):
        demo.demo_textual(Image)
        async with cast(AsyncContextManager[Pilot[Any]], awaitable) as pilot:
            pilot.app.query_one(Button).press()
            await pilot.pause()
