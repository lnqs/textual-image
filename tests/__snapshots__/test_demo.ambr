# serializer version: 1
# name: test_rich
  '''
  KittyImage from path:
         ░▒ ░ ▒▓░░░░░░░░░░░░░░░▒▒
   ░      ░ ░▒▓░▒█▓░▒▓▒▒▒░░░░░░▒▒
  ░▓ ▒▓ ░░░ ░░░░▒▒▒▒▒▒▒░░▒▒▒▒▒▒▒▒
     ░░ ░░░             ░▒▒▒▒▒▒▒▒
  ░░      ░  ▒▒▒▓▓▓▓▒▒░ ░▒▒▒▒▒▒▒▒
  ▒▒▒▒  ░ ░░░░▒░▒▓▓▒░▒▒░▒▒▒▒▒▒▒▒▒
          ░▒▒▒▒▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒
  ░      ▒▓▓▓▒▓▒▒▒▒▒▒▒▓▓▒▒▒▒▒▒▒▒▒
  ▒     ░▓▓▓▒░▒▒▒▒▒▒▒▓▒▒▒▒▒▒▒▒▒▒▒
         ▒▓▒▒░▒▒▒▒▒▒▒▒░▒▒▒▒▒▒▒▒▒▒
    ░▒░   ░░  ▒▒░░░░▒▓  ░░▒▒▒▒▒▒▒
    ▒░        ▓▓▒▒▒▒█▒        ░▒▒
  ▓           ░█▒░░▒▓           ░
  ▓▒             ▒▓░             
  ▓░   ▒░░░░    ▓█▒              
  ▓    ░░░░░    ░░    ░░▒▒░░░    
  ▒                              
                ░░               
  
  KittyImage from pillow image:
         ░▒ ░ ▒▓░░░░░░░░░░░░░░░▒▒
   ░      ░ ░▒▓░▒█▓░▒▓▒▒▒░░░░░░▒▒
  ░▓ ▒▓ ░░░ ░░░░▒▒▒▒▒▒▒░░▒▒▒▒▒▒▒▒
     ░░ ░░░             ░▒▒▒▒▒▒▒▒
  ░░      ░  ▒▒▒▓▓▓▓▒▒░ ░▒▒▒▒▒▒▒▒
  ▒▒▒▒  ░ ░░░░▒░▒▓▓▒░▒▒░▒▒▒▒▒▒▒▒▒
          ░▒▒▒▒▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒
  ░      ▒▓▓▓▒▓▒▒▒▒▒▒▒▓▓▒▒▒▒▒▒▒▒▒
  ▒     ░▓▓▓▒░▒▒▒▒▒▒▒▓▒▒▒▒▒▒▒▒▒▒▒
         ▒▓▒▒░▒▒▒▒▒▒▒▒░▒▒▒▒▒▒▒▒▒▒
    ░▒░   ░░  ▒▒░░░░▒▓  ░░▒▒▒▒▒▒▒
    ▒░        ▓▓▒▒▒▒█▒        ░▒▒
  ▓           ░█▒░░▒▓           ░
  ▓▒             ▒▓░             
  ▓░   ▒░░░░    ▓█▒              
  ▓    ░░░░░    ░░    ░░▒▒░░░    
  ▒                              
                ░░               
  
  Same KittyImage instance rendered twice:
         ░▒ ░ ▒▓░░░░░░░░░░░░░░░▒▒
   ░      ░ ░▒▓░▒█▓░▒▓▒▒▒░░░░░░▒▒
  ░▓ ▒▓ ░░░ ░░░░▒▒▒▒▒▒▒░░▒▒▒▒▒▒▒▒
     ░░ ░░░             ░▒▒▒▒▒▒▒▒
  ░░      ░  ▒▒▒▓▓▓▓▒▒░ ░▒▒▒▒▒▒▒▒
  ▒▒▒▒  ░ ░░░░▒░▒▓▓▒░▒▒░▒▒▒▒▒▒▒▒▒
          ░▒▒▒▒▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒
  ░      ▒▓▓▓▒▓▒▒▒▒▒▒▒▓▓▒▒▒▒▒▒▒▒▒
  ▒     ░▓▓▓▒░▒▒▒▒▒▒▒▓▒▒▒▒▒▒▒▒▒▒▒
         ▒▓▒▒░▒▒▒▒▒▒▒▒░▒▒▒▒▒▒▒▒▒▒
    ░▒░   ░░  ▒▒░░░░▒▓  ░░▒▒▒▒▒▒▒
    ▒░        ▓▓▒▒▒▒█▒        ░▒▒
  ▓           ░█▒░░▒▓           ░
  ▓▒             ▒▓░             
  ▓░   ▒░░░░    ▓█▒              
  ▓    ░░░░░    ░░    ░░▒▒░░░    
  ▒                              
                ░░               
         ░▒ ░ ▒▓░░░░░░░░░░░░░░░▒▒
   ░      ░ ░▒▓░▒█▓░▒▓▒▒▒░░░░░░▒▒
  ░▓ ▒▓ ░░░ ░░░░▒▒▒▒▒▒▒░░▒▒▒▒▒▒▒▒
     ░░ ░░░             ░▒▒▒▒▒▒▒▒
  ░░      ░  ▒▒▒▓▓▓▓▒▒░ ░▒▒▒▒▒▒▒▒
  ▒▒▒▒  ░ ░░░░▒░▒▓▓▒░▒▒░▒▒▒▒▒▒▒▒▒
          ░▒▒▒▒▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒
  ░      ▒▓▓▓▒▓▒▒▒▒▒▒▒▓▓▒▒▒▒▒▒▒▒▒
  ▒     ░▓▓▓▒░▒▒▒▒▒▒▒▓▒▒▒▒▒▒▒▒▒▒▒
         ▒▓▒▒░▒▒▒▒▒▒▒▒░▒▒▒▒▒▒▒▒▒▒
    ░▒░   ░░  ▒▒░░░░▒▓  ░░▒▒▒▒▒▒▒
    ▒░        ▓▓▒▒▒▒█▒        ░▒▒
  ▓           ░█▒░░▒▓           ░
  ▓▒             ▒▓░             
  ▓░   ▒░░░░    ▓█▒              
  ▓    ░░░░░    ░░    ░░▒▒░░░    
  ▒                              
                ░░               
  
  KittyImage with width set to 10:
  ░░░░▒▒░░▒▒
  ░ ░░▒▒▒▒▒▒
    ▒▒▒▒▒▒▒▒
  ░   ▒▒   ░
            
  
  KittyImage inside table:
  ┏━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
  ┃ Name         ┃ Birthday         ┃ Picture                         ┃
  ┡━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
  │ Grace Hopper │ December 9, 1906 │        ░▒ ░ ▒▓░░░░░░░░░░░░░░░▒▒ │
  │              │                  │  ░      ░ ░▒▓░▒█▓░▒▓▒▒▒░░░░░░▒▒ │
  │              │                  │ ░▓ ▒▓ ░░░ ░░░░▒▒▒▒▒▒▒░░▒▒▒▒▒▒▒▒ │
  │              │                  │    ░░ ░░░             ░▒▒▒▒▒▒▒▒ │
  │              │                  │ ░░      ░  ▒▒▒▓▓▓▓▒▒░ ░▒▒▒▒▒▒▒▒ │
  │              │                  │ ▒▒▒▒  ░ ░░░░▒░▒▓▓▒░▒▒░▒▒▒▒▒▒▒▒▒ │
  │              │                  │         ░▒▒▒▒▓▓▓▓▓▓▓▒▒▒▒▒▒▒▒▒▒▒ │
  │              │                  │ ░      ▒▓▓▓▒▓▒▒▒▒▒▒▒▓▓▒▒▒▒▒▒▒▒▒ │
  │              │                  │ ▒     ░▓▓▓▒░▒▒▒▒▒▒▒▓▒▒▒▒▒▒▒▒▒▒▒ │
  │              │                  │        ▒▓▒▒░▒▒▒▒▒▒▒▒░▒▒▒▒▒▒▒▒▒▒ │
  │              │                  │   ░▒░   ░░  ▒▒░░░░▒▓  ░░▒▒▒▒▒▒▒ │
  │              │                  │   ▒░        ▓▓▒▒▒▒█▒        ░▒▒ │
  │              │                  │ ▓           ░█▒░░▒▓           ░ │
  │              │                  │ ▓▒             ▒▓░              │
  │              │                  │ ▓░   ▒░░░░    ▓█▒               │
  │              │                  │ ▓    ░░░░░    ░░    ░░▒▒░░░     │
  │              │                  │ ▒                               │
  │              │                  │               ░░                │
  └──────────────┴──────────────────┴─────────────────────────────────┘
  [3;35mNone[0m
  
  '''
# ---
